quarkus/tests/integration/src/main/java/org/keycloak/it/utils/KeycloakDistribution.java

```diff
@@ -0,0 +1,337 @@
+/*
+ * Copyright 2021 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.it.utils;
+
+import static org.keycloak.quarkus.runtime.Environment.LAUNCH_MODE;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.net.HttpURLConnection;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.security.KeyManagementException;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.security.cert.X509Certificate;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+import javax.net.ssl.HostnameVerifier;
+import javax.net.ssl.HttpsURLConnection;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.SSLSession;
+import javax.net.ssl.SSLSocketFactory;
+import javax.net.ssl.TrustManager;
+import javax.net.ssl.X509TrustManager;
+import org.apache.commons.io.FileUtils;
+import org.eclipse.aether.artifact.Artifact;
+import org.jboss.logging.Logger;
+
+import io.quarkus.bootstrap.util.ZipUtils;
+
+public final class KeycloakDistribution {
+
+    private static final Logger LOGGER = Logger.getLogger(KeycloakDistribution.class);
+
+    private Process keycloak;
+    private int exitCode = -1;
+    private final Path distPath;
+    private final List<String> outputStream = new ArrayList<>();
+    private final List<String> errorStream = new ArrayList<>();
+    private boolean reCreate;
+    private boolean manualStop;
+    private String relativePath;
+    private int httpPort;
+    private boolean debug;
+    private ExecutorService outputExecutor;
+
+    public <T> KeycloakDistribution() {
+        distPath = prepareDistribution();
+    }
+
+    public void start(List<String> arguments) {
```

Comment:
All this logic can be immensely simplified using Testcontainers with the Docker image we build here:
https://github.com/keycloak/keycloak-containers/pull/340

Reply:
Yes, it can. As I mentioned in the description we should be able to support this scenario very easily now by having a specific implementation for running it as a container.

But I don't think that excludes running/testing the server without a container.

Reply:
> But I don't think that excludes running/testing the server without a container.

I agree it's a matter of tradeoffs 🙂 
With Docker:
- reuse of the Testcontainers framework for (among the others) readiness/log management/dangling resources
- will test the `Dockerfile` itself as well
- more isolated environments (e.g. can use different Java versions for server / cli)

Without Docker:
- (probably) faster tests execution
- more integrated workflow

Reply:
Yeah ... I think your proposal fits nicely once we start playing around with dev services.